From 696053884954cace50bac2879f909d65a13fcfed Mon Sep 17 00:00:00 2001
From: Simon Hausmann <simon.hausmann@theqtcompany.com>
Date: Tue, 5 May 2015 10:52:34 +0200
Subject: [PATCH] Fix memory corruption in array handling

SimpleArrayData's markObjects() implementation did not mark the entries
correctly. When the dequeue offset was non-zero, we would end up marking values
that may have been garbage collected earlier.

Task-number: QTBUG-45888
Change-Id: Iacec350ccc76399ad4d16138af50acf22b2809db
---
 src/qml/jsruntime/qv4arraydata.cpp | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/src/qml/jsruntime/qv4arraydata.cpp b/src/qml/jsruntime/qv4arraydata.cpp
index 737c891..d34636d 100644
--- a/src/qml/jsruntime/qv4arraydata.cpp
+++ b/src/qml/jsruntime/qv4arraydata.cpp
@@ -216,8 +216,10 @@ void ArrayData::ensureAttributes(Object *o)
 void SimpleArrayData::markObjects(Heap::Base *d, ExecutionEngine *e)
 {
     Heap::SimpleArrayData *dd = static_cast<Heap::SimpleArrayData *>(d);
-    uint l = dd->len;
-    for (uint i = 0; i < l; ++i)
+    uint remainingEntriesToMark = dd->len;
+    for (uint i = dd->offset; i < dd->alloc && remainingEntriesToMark > 0; ++i, --remainingEntriesToMark)
+        dd->arrayData[i].mark(e);
+    for (uint i = 0; i < dd->offset && remainingEntriesToMark > 0; ++i, --remainingEntriesToMark)
         dd->arrayData[i].mark(e);
 }
 
-- 
1.9.0

